spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Lars.scala

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.optim
+
+import com.intel.analytics.bigdl.Module
+import com.intel.analytics.bigdl.nn.Container
+import com.intel.analytics.bigdl.nn.abstractnn.{AbstractModule, Activity}
+import com.intel.analytics.bigdl.optim.SGD.{Default, LearningRateSchedule}
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.Table
+import org.apache.log4j.{Level, Logger}
+
+import scala.reflect.ClassTag
+
+class Lars[@specialized(Float, Double) T: ClassTag](
+                                                     lrScheduleOwner: Boolean,
+                                                     trust: Double = 1.0,
+                                                     learningRate: Double = 1e-3,
+                                                     learningRateDecay: Double = 0.01,
+                                                     weightDecay: Double = 0.0005,
+                                                     momentum: Double = 0.5,
+                                                     learningRateSchedule: LearningRateSchedule = Default()
+                                                   )(implicit ev: TensorNumeric[T])
+  extends OptimMethod[T]
+{
+  @transient
+  private var buffer: Tensor[T] = null
+  private val sgd = new SGD[T](learningRate,learningRateDecay,weightDecay,momentum,learningRateSchedule=this.learningRateSchedule)
```

Comment:
why not make LARS extending SGD? Probably name it as LarsSGD ?

Reply:
SGD[T] is a specialized class and extending it will trigger a scala compiler issue (not sure if it is a bug). So it’s a workaround here to have a SDG instance in Lars. The scala compiler issue [reference](https://github.com/scala/bug/issues/8405)